### ISSUE TITLE
[typescript-estree] jsx option doesn't work if project option exists

### ISSUE THREAD

**Issue by mysticatea** [typescript-eslint/typescript-eslint#566]
**What code were you trying to parse?**

```ts
a = <div/>
```

With `jsx` option.

```js
const estree = require("@typescript-eslint/typescript-estree")
const code = `a = <div/>`

try {
    estree.parseAndGenerateServices(code, {
        filePath: "test.js",
        jsx: false,
        project: "./tsconfig.json"
    })
} catch (error) {
    console.log(error)
}
```

**What did you expect to happen?**

It throws a syntax error: `{ index: 8, lineNumber: 1, column: 8, message: "'>' expected." }`

**What actually happened?**

It succeeded to parse.

If I removed `project:"./tsconfig.json"` from the option, it works expected.

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `1.9.0` |
| `TypeScript`                           | `3.4.5` |
| `node`                                 | `12.2.0` |
| `npm`                                  | `6.9.0` |

---

This is confusing ESLint that lints non-ts files.
For example, it doesn't respect `parserOptions.ecmaFeatures.jsx:true` for TypeScript inside Vue files if `parserOptions.project` existed.

```js
const estree = require("@typescript-eslint/typescript-estree")
const code = `a = <div/>`

estree.parseAndGenerateServices(code, { // throws syntax error due to not respect `jsx: true` option 
    filePath: "test.vue",
    jsx: true,
    project: "./tsconfig.json"
})
```

**Comment by stuartZhang** [comment]
Are there any workarounds for the issue? I'd like to use ESLint + typesccript in my project.

**Comment by bradzacher** [comment]
@stuartZhang - there is no need to have a workaround for your projects using eslint.
As long as you respect the `x` suffix on extensions, everything will work as expected.
I.e. name your files `.tsx`/`.jsx`, or else the typescript compiler (`tsc`) will hard fail if you don't (see #574).

----

This issue is in regards to the programmatic usage of `typescript-estree`.
@mysticatea - I don't think this is something we can fix without being really hacky; the typescript parser does not allow JSX in files that aren't.

The option is only there if you chose not to pass in a filename IIRC.

**Comment by mysticatea** [comment]
@bradzacher I don't think so because it's works fine if project option doesn't exist. We can use the same way.

**Comment by bradzacher** [comment]
There are two separate code paths for parsing depending on if you have a project or not.

If you don't have a project, we essentially just invoke the parser like this (pseudocode, not actual invocation - [this is the actual invocation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts#L101-L152)):
```ts
function doParse(fileContent: string, options: ParserOptions) {
  const FILENAME = options.jsx ? 'fakefilename.tsx' : 'fakefilename.ts';
  const ast = ts.parser({
    files: [FILENAME],
    compilerOptions: {
      isolatedModules: true,
      jsx: options.jsx ? 'preserve' : undefined,
    },
    compilerHost: {
      getFileContent(filename: string) {
        // there's only one file in this fake project
        return fileContent;
      },
    },
  });
}
```
We can do this because _just_ parsing is cheap, so we don't need TS to store anything about the file - meaning we can give it a fake filename with an extension which matches your setting.

---

HOWEVER, when you pass `project` in, it's a different story, we have to do an entire project parse to provide correct types. This is important because typescript is the one that coordinates that. It is the one that reads the filesystem, and parses every file referenced by the `tsconfig.json` - it's essentially pre-parsing every single file.

We do some magic (using internal tsc things that are hidden from the types by an `/* @internal */` comment in the TS codebase) to make sure that we're providing the correct file content where applicable, but we don't spoof filenames because typescript is the one controlling that.

(I believe [this is the code](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/tsconfig-parser.ts#L67-L186) that sets this up).

@JamesHenry or @uniqueiniquity can correct me if I'm wrong in there as they're SMEs in that area.

**Comment by mysticatea** [comment]
Thank you for elaboration.

In that case, if we cannot overwrite `compilerOptions.jsx`, should we remove `jsx` option? The option is very confusing us.

**Comment by bradzacher** [comment]
It's definitely something I've thought about considering we can't fully respect it.

I don't see why not, apart from the fact that it "breaks" compat with eslint.
The parser is outside of my explicit area of expertise though, so I'd be looking to @JamesHenry and @uniqueiniquity to answer that.

**Comment by edvardchen** [comment]
I encountered this issue when writing test cases for my own rule and found this discussion. Thanks everyone above and I want to share my solution here for writing eslint rule test cases. 

```javascript
const tsTester = new RuleTester({
  parser: '@typescript-eslint/parser',
  parserOptions: {
    sourceType: 'module',
    project: path.resolve(__dirname, 'tsconfig.json')
  }
});

tsTester.run('your_rule', rule, {
  // specify the filename with tsx extension 
  valid: [{ code: '<div />', filename: 'a.tsx' }],
  invalid: []
});
```
Hope it helpful

**Comment by bradzacher** [comment]
Docs and handling clarified in 366518f75944dd387d2f55a33a31d2ac9c743f1d
